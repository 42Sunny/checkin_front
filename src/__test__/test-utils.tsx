// test-utils.jsx
import { configureStore } from "@reduxjs/toolkit";
import { Queries, render, RenderResult } from "@testing-library/react";
import { createMemoryHistory } from "history";
import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import { Router } from "react-router-dom";
// Import your own reducer
import renderer from "react-test-renderer";
import appStore from "../redux/configureStore";
import rootReducer, { RootState } from "../redux/modules";
import { initialState as clusterInitState } from "../redux/modules/cluster";
import { initialState as userInitState } from "../redux/modules/user";

const makeCustomStore = ({
  cluster,
  user,
}: {
  cluster?: typeof clusterInitState;
  user?: typeof userInitState;
} = {}) => {
  return {
    clusterReducer: { ...clusterInitState, ...cluster },
    userReducer: { ...userInitState, ...user },
  };
};

const customRender = (
  ui: ReactElement,
  preloadedState: RootState = makeCustomStore(),
  { store = configureStore({ reducer: rootReducer, preloadedState }), ...renderOptions } = {},
  history = createMemoryHistory(),
): RenderResult<Queries, HTMLElement> => {
  const Wrapper: any = ({ children }: { children: ReactElement }) => {
    return (
      <Provider store={store}>
        <Router history={history}>{children}</Router>
      </Provider>
    );
  };
  return render(ui, { wrapper: Wrapper, ...renderOptions });
};

const createWithWrapper = (children: ReactElement) =>
  renderer.create(
    <Provider store={appStore}>
      <Router history={createMemoryHistory()}>{children}</Router>
    </Provider>,
  );

// re-export everything
export * from "@testing-library/react";
// override render method
export { customRender as render, createWithWrapper, makeCustomStore };
