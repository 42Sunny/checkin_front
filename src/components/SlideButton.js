import React, { useCallback, useEffect } from "react";
import "../styles/Slider.css";
const SlideButton = ({ value, setValue }) => {
  const setSlideValue = useCallback(
    (e) => {
      const va = e.currentTarget.value;
      setValue(va);
    },
    [setValue]
  );

  useEffect(() => {
    const slider = document.getElementById("slider");
    const sliderText = document.querySelector(".slider-background-text");

    const checkSliderValue = (e) => {
      const currentValue = e.target.value;

      if (currentValue < 80) {
        // slider.style.background = "white";
        sliderText.style.opacity=.5
        setValue(0);
      } else setValue(100);
    };

    const changeSlider = (e) => {
      const currentValue = e.target.value;
      if (currentValue > 20) sliderText.style.opacity= 0;
      if (currentValue > 80) slider.style.background = "rgba(211, 211, 211, 1)";
      else slider.style.background = "";
    };
    slider.addEventListener("mouseup", checkSliderValue);
    slider.addEventListener("touchend", checkSliderValue);
    slider.addEventListener("mousemove", changeSlider);
    slider.addEventListener("touchmove", changeSlider);
    return () => {
      setValue(0);
      slider.removeEventListener("mouseup", checkSliderValue);
      slider.removeEventListener("touchend", checkSliderValue);
    };
  }, [setValue]);

  return (
    <div className="slider-container">
      <input placeholder="slide to quit" type="range" value={value} min={1} max={100} onChange={setSlideValue} id="slider"></input>
      <div className="slider-background-text">밀어서 체크아웃</div>
    </div>
  );
};

export default SlideButton;
