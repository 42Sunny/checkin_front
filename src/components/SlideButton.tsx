import React, { useLayoutEffect, useRef } from "react";
import classes from "../styles/SlideButton.module.css";

interface IProps {
  value: number;
  setValue: React.Dispatch<React.SetStateAction<number>>;
}

const SlideButton: React.FC<IProps> = ({ value, setValue }) => {
  const slider = useRef<HTMLInputElement>(null);
  const sliderText = useRef<HTMLParagraphElement>(null);

  const handleSliderChange = (
    e:
      | React.MouseEvent<HTMLInputElement>
      | React.TouchEvent<HTMLInputElement>
      | React.ChangeEvent<HTMLInputElement>,
  ) => {
    if (e.currentTarget.value === "100") setValue(99);
    else setValue(parseInt(e.currentTarget.value, 10));
  };

  const handleSliderTouchEnd = (
    e: React.MouseEvent<HTMLInputElement> | React.TouchEvent<HTMLInputElement>,
  ) => {
    if (!(slider.current && sliderText.current)) return;
    const currentValue = parseInt(e.currentTarget.value, 10);
    if (currentValue < 80) setValue(0);
    else setValue(100);
  };

  useLayoutEffect(() => {
    if (!(slider.current && sliderText.current)) return;

    if (value > 1) sliderText.current.style.opacity = "0";
    else sliderText.current.style.opacity = "";

    if (value > 80) slider.current.style.background = "black";
    else slider.current.style.background = "";
  }, [value]);

  return (
    <>
      <div className={classes["slider-wrapper"]}>
        <input
          onMouseUp={handleSliderTouchEnd}
          onTouchEnd={handleSliderTouchEnd}
          onMouseMove={handleSliderChange}
          onTouchMove={handleSliderChange}
          ref={slider}
          type='range'
          value={value}
          min={1}
          max={100}
          onChange={handleSliderChange}
          className={classes.slider}
        />
        <p ref={sliderText} className={classes["slider-backgroundText"]}>
          밀어서 체크아웃
        </p>
      </div>
    </>
  );
};

export default SlideButton;
