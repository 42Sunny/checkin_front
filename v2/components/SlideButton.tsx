import React, { useEffect, useLayoutEffect, useRef, useState } from "react";
import classes from "../styles/components/SlideButton.module.css";

interface IProps {
  value: number;
  setValue: React.Dispatch<React.SetStateAction<number>>;
}

const SlideButton: React.FC<IProps> = ({ value, setValue }) => {
  const slider = useRef<HTMLInputElement>(null);
  const [isResetted, setIsResetted] = useState(false);

  const handleSliderChange = (
    e:
      | React.MouseEvent<HTMLInputElement>
      | React.TouchEvent<HTMLInputElement>
      | React.ChangeEvent<HTMLInputElement>,
  ) => {
    setValue(e.currentTarget.valueAsNumber);
  };

  const handleSliderTouchEnd = (
    e: React.MouseEvent<HTMLInputElement> | React.TouchEvent<HTMLInputElement>,
  ) => {
    if (!slider.current) return;
    const currentValue = e.currentTarget.valueAsNumber;
    if (currentValue < 80) {
      setIsResetted(true);
    } else {
      setValue(100);
    }
  };

  useLayoutEffect(() => {
    if (!slider.current) return;
    if (value > 80) {
      slider.current.style.backgroundColor = "black";
    } else slider.current.style.background = "";
  }, [value]);

  useEffect(() => {
    if (isResetted) {
      setValue(0);
      setIsResetted(false);
    }
  }, [isResetted, setValue, value]);

  return (
    <div className={classes["slider-wrapper"]}>
      <input
        onMouseUp={handleSliderTouchEnd}
        onTouchEnd={handleSliderTouchEnd}
        ref={slider}
        type='range'
        value={value}
        min={0}
        max={99}
        onInput={handleSliderChange}
        className={classes.slider}
      />
      {value === 0 && (
        <p role='contentinfo' className={classes["slider-backgroundText"]}>
          밀어서 체크아웃
        </p>
      )}
    </div>
  );
};

export default SlideButton;
