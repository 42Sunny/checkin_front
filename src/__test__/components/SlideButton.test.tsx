import { fireEvent } from "@testing-library/react";
import SlideButton from "../../components/SlideButton";
import { createWithWrapper, render, screen } from "../test-utils";

describe("src/components/SlideButton.tsx", () => {
  const slideButtonProps = {
    value: 0,
    setValue: () => {},
  };
  beforeEach(() => {
    slideButtonProps.setValue = jest.fn();
    slideButtonProps.value = 0;
  });
  it("matches snapshot", () => {
    const component = createWithWrapper(<SlideButton {...slideButtonProps} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("user drags slide button to end", () => {
    render(<SlideButton {...slideButtonProps} />);
    const slideBtnEl = screen.getByRole("slider");

    // slide이벤트가 jsdom에 따로 없어서 input으로 대체함
    fireEvent.input(slideBtnEl, { target: { value: 99 } });
    expect(slideButtonProps.setValue).toBeCalledWith(99);
    jest.resetAllMocks();
    fireEvent.touchEnd(slideBtnEl, { target: { value: 99 } });
    expect(slideButtonProps.setValue).toBeCalledWith(100);
  });

  it("user drags slide button to 80", () => {
    render(<SlideButton {...slideButtonProps} />);
    const slideBtnEl = screen.getByRole("slider");

    fireEvent.input(slideBtnEl, { target: { value: "80" } });
    expect(slideButtonProps.setValue).toBeCalledWith(80);
    jest.resetAllMocks();
    fireEvent.touchEnd(slideBtnEl, { target: { value: "80" } });
    expect(slideButtonProps.setValue).toBeCalledWith(100);
  });

  it("user drags slide button to 50 and slide button resets its value", () => {
    render(<SlideButton {...slideButtonProps} />);
    const slideBtnEl = screen.getByRole("slider");

    fireEvent.input(slideBtnEl, { target: { value: 50 } });
    expect(slideButtonProps.setValue).toBeCalledWith(50);
    jest.resetAllMocks();
    fireEvent.touchEnd(slideBtnEl, { target: { value: 50 } });
    expect(slideButtonProps.setValue).toBeCalledWith(0);
  });

  it("user drags slide button to 20 and slide button resets its value)", () => {
    render(<SlideButton {...slideButtonProps} />);
    const slideBtnEl = screen.getByRole("slider");

    fireEvent.input(slideBtnEl, { target: { value: 20 } });
    expect(slideButtonProps.setValue).toBeCalledWith(20);
    jest.resetAllMocks();
    fireEvent.touchEnd(slideBtnEl, { target: { value: 20 } });
    expect(slideButtonProps.setValue).toBeCalledWith(0);
  });

  it("slide button does render text when value is 0", () => {
    render(<SlideButton {...slideButtonProps} value={0} />);

    const slideTextEl = screen.queryByText(/밀어서 체크아웃/i);

    expect(slideTextEl).toBeInTheDocument();
  });

  it("slide button does render text when value is above 0", () => {
    render(<SlideButton {...slideButtonProps} value={1} />);

    const slideTextEl = screen.queryByText(/밀어서 체크아웃/i);

    expect(slideTextEl).not.toBeInTheDocument();
  });

  it("slide button doesn't have background black when value is under 80", () => {
    render(<SlideButton {...slideButtonProps} value={0} />);

    const slideBtnEl = screen.getByRole("slider");

    expect(slideBtnEl.style.backgroundColor).not.toBe("black");
  });

  it("slide button has background black when value is above 80", () => {
    render(<SlideButton {...slideButtonProps} value={81} />);

    const slideBtnEl = screen.getByRole("slider");

    expect(slideBtnEl.style.backgroundColor).toBe("black");
  });
});
