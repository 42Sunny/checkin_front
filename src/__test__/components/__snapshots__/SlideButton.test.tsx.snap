// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/SlideButton.tsx matches snapshot 1`] = `
<div
  className="slider-wrapper"
>
  <input
    className="slider"
    max={99}
    min={0}
    onInput={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    type="range"
    value={0}
  />
  <p
    className="slider-backgroundText"
    role="contentinfo"
  >
    밀어서 체크아웃
  </p>
</div>
`;
